nixos/modules/services/databases/kanidm.nix

```diff
@@ -0,0 +1,101 @@
+{ config, pkgs, lib, ... }:
+let
+  cfg = config.services.kanidm;
+  settingsFormat = pkgs.formats.toml { };
+in
+{
+
+  options.services.kanidm = {
+    enable = lib.mkEnableOption "kanidm service";
+
+    ensureDomainName = lib.mkOption {
+      description = ''
+        The domain_name that Kanidm manages. It must be a descendant or equal to the domain
+        specified in settings.origin. It is always set when the server starts up.
+        While in theory it could be changed later, you should carefully consider these
+        warnings:
+        <link xlink:href="https://kanidm.github.io/kanidm/administrivia.html#rename-the-domain"/>
+      '';
+      example = "example.org";
+      type = lib.types.str;
+    };
+
+    settings = lib.mkOption {
+      type = lib.types.submodule {
+
+        freeformType = settingsFormat.type;
+
+        options.bindaddress = lib.mkOption {
+          description = "Address/port combination the webserver binds to.";
+          example = "[::1]:8443";
+          type = lib.types.str;
+        };
+        options.origin = lib.mkOption {
+          description = "The origin of your kanidm instance. Must have https as protocol.";
+          example = "https://idm.example.org";
+          type = lib.types.str;
+        };
+        options.db_path = lib.mkOption {
+          description = "Path to kanidm database.";
+          default = "/var/lib/kanidm/kanidm.db";
+          type = lib.types.path;
+        };
+        options.log_level = lib.mkOption {
+          description = "Log level of the server.";
+          default = "default";
+          type = lib.types.enum [ "default" "verbose" "perfbasic" "perffull" ];
+        };
+        options.role = lib.mkOption {
+          description = "The role of this server. This affects features available and how replication may interact.";
+          default = "WriteReplica";
+          type = lib.types.enum [ "WriteReplica" "WriteReplicaNoUI" "ReadOnlyReplica" ];
+        };
+
+      };
+      default = { };
+      description = ''
+        Settings for kanidm, see
+        <link xlink:href="https://github.com/kanidm/kanidm/blob/master/examples/server.toml"/>
+        for example config.
+      '';
+    };
+  };
+
+  config =
+    lib.mkIf cfg.enable {
+
+      systemd.services.kanidm = {
+        description = "Kanidm server";
+        after = [ "network.target" ];
+        wantedBy = [ "multi-user.target" ];
+        serviceConfig = {
+          ExecPreStart = ''
+            ${pkgs.kanidm}/bin/kanidmd domain_name_change \
+              --config ${settingsFormat.generate "config.toml" cfg.settings} \
+              -n ${cfg.ensureDomainName}
+          '';
```

Comment:
This seems to be a pretty expensive operation and according to the kanidm docs
> WARNING: This WILL break ALL u2f/webauthn tokens that have been enrolled, which MAY cause accounts to be locked out and unrecoverable until further action is taken. DO NOT CHANGE the domain_name unless REQUIRED and have a plan on how to manage these issues.

So, I think if we add support to change the domain name in nix, we only want to run this if the domain is different.

Reply:
I assumed naively, that it doesn't have a cost at all if the name isn't changed? At least thats what I'd expect from the implementation. I didn't find a way to fetch the current domain_name from the db. If there is one we could introduce a check.

Reply:
I briefly looked at the implementation but I didn’t see a check to do nothing if there is no change. We could store the domain in a text file near the db file and only run this if it changed.

Reply:
Maybe we could ask upstream to implement either such a check or a possibilty to retrieve the domain name by querying the domain? Adding a text file would work, but I'm somewhat hesitant to add it as a de-facto config file, considering that upstream explicitly doesn't want to have `domain_name` as config, but as a database property:

https://github.com/kanidm/kanidm/issues/616
https://github.com/kanidm/kanidm/pull/617#issuecomment-978350062

Do you agree in principle that the `ensureDomainName` should be a thing?

Reply:
I’d like it if things get more declarative. Both of your ideas sound better than having a text file :)